src/goto-instrument/cover.cpp

```diff
@@ -165,91 +163,53 @@ void parse_cover_options(const cmdlinet &cmdline, optionst &options)
     cmdline.isset("cover-traces-must-terminate"));
 }
 
-/// Applies instrumenters to given goto functions
-/// \param goto_functions: the goto functions
-/// \param instrumenters: the instrumenters
-/// \param function_filters: function filters to discard certain functions
-/// \param message_handler: a message handler
-void instrument_cover_goals(
-  goto_functionst &goto_functions,
-  const cover_instrumenterst &instrumenters,
-  const function_filterst &function_filters,
-  message_handlert &message_handler)
-{
-  Forall_goto_functions(f_it, goto_functions)
-  {
-    if(!function_filters(f_it->first, f_it->second))
-      continue;
-
-    instrument_cover_goals(f_it->second.body, instrumenters, message_handler);
-  }
-}
-
-/// Instruments goto functions based on given command line options
-/// \param options: the options
-/// \param symbol_table: the symbol table
-/// \param goto_functions: the goto functions
-/// \param message_handler: a message handler
-bool instrument_cover_goals(
+/// Build data structures controlling coverage from command-line options.
+/// \param options: command-line options
+/// \param symbol_table: global symbol table
+/// \param message_handler: used to log incorrect option specifications
+/// \return a cover_configt on success, or optionalt() otherwise.
+std::unique_ptr<cover_configt> get_cover_config(
   const optionst &options,
   const symbol_tablet &symbol_table,
-  goto_functionst &goto_functions,
   message_handlert &message_handler)
 {
   messaget msg(message_handler);
+  std::unique_ptr<cover_configt> config = util_make_unique<cover_configt>();
```

Comment:
Why does this need to be a pointer?

Reply:
I'd rather not implement move constructors for all the filter kinds -- simpler to let them not support copy/move and pass a handle to the config object instead. Unique-ptr really takes the pain out of managing its lifespan in any case.

Reply:
It just seems a lot of machinery for a very simple problem where performance doesn't matter either. Why not just return `cover_configt` by value?

Reply:
Same thing -- needing copyable and movable everything in the config object. So much simpler to just allocate one and pass it around.

Reply:
Ok. So while I _would_ have done it differently, you actually _did it_, so you win :-)

Reply:
Are you happy to approve in that case? This is still blocked on code-owner approval